virtcontainers/device/drivers/vfio.go

```diff
@@ -97,17 +100,61 @@ func (device *VFIODevice) DeviceType() config.DeviceType {
 	return device.DevType
 }
 
+type VFIODevice_Type int32
+
+const (
+        VFIODEVICE_TYPE_VFIO_ERROR    VFIODevice_Type = 0
+        VFIODEVICE_TYPE_VFIO_NORMAL   VFIODevice_Type = 1
+        VFIODEVICE_TYPE_VFIO_MEDIATED VFIODevice_Type = 2
+)
+
+func getVFIODetails(deviceFileName, iommuDevicesPath string) (deviceBDF, deviceSysfsDev string, vfioDeviceType VFIODevice_Type, err error) {
+
+        tokens := strings.Split(deviceFileName, ":")
+        vfioDeviceType = VFIODEVICE_TYPE_VFIO_ERROR
+        if len(tokens) == 3 {
+                vfioDeviceType = VFIODEVICE_TYPE_VFIO_NORMAL
+        } else {
+                tokens = strings.Split(deviceFileName, "-")
+                if len(tokens) == 5 {
+                        vfioDeviceType = VFIODEVICE_TYPE_VFIO_MEDIATED
+                }
+        }
+
+        err = nil
+        deviceBDF = ""
+        deviceSysfsDev = ""
+        if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_NORMAL) {
+                //Get bdf of device eg 0000:00:1c.0
+                deviceBDF = getBDF(deviceFileName)
+        } else if (vfioDeviceType == VFIODEVICE_TYPE_VFIO_MEDIATED){
+                //Get sysfsdev of device eg f79944e4-5a3d-11e8-99ce-479cbab002e4
+                sysfsDevStr := filepath.Join(iommuDevicesPath, deviceFileName)
+                deviceSysfsDev, err = getSysfsDev(sysfsDevStr)
+        } else {
+                err = fmt.Errorf("Incorrect tokens found while parsing vfio details: %s", deviceFileName)
+        }
+
+        if err != nil {
+                return "", "", vfioDeviceType, err;
+        } else {
```

Comment:
this `else` is not necessary

Reply:
As I am NOT sure about whether nil will be definitely returned or NOT when error is encountered on calling filepath.EvalSymlinks(), so put this case into if(...)